src/vs/workbench/contrib/terminal/browser/terminalProcessManager.ts

```diff
@@ -328,7 +333,20 @@ export class TerminalProcessManager extends Disposable implements ITerminalProce
 		const useConpty = this._configHelper.config.windowsEnableConpty && !isScreenReaderModeEnabled;
 		const shouldPersist = this._configHelper.config.enablePersistentSessions && !shellLaunchConfig.isFeatureTerminal;
 
-		this._terminalInstanceService.onPtyHostUnresponsive(() => this._onPtyDisconnect.fire());
+		this._register(this._terminalInstanceService.onPtyHostUnresponsive(() => {
+			this.isDisconnected = true;
+			this._onPtyDisconnect.fire();
+		}));
+		this._ptyResponsiveListener = this._terminalInstanceService.onPtyHostResponsive(() => {
+			this.isDisconnected = false;
+			this._onPtyReconnect.fire();
+		});
+		this._register(toDisposable(() => this._ptyResponsiveListener?.dispose()));
+		this._register(this._terminalInstanceService.onPtyHostRestart(() => {
+			// When the pty host restarts, reconnect is no longer possible
+			this._ptyResponsiveListener?.dispose();
+			this._ptyResponsiveListener = undefined;
```

Comment:
I don't see where we check if `this._ptyResponsiveListener` is defined

Reply:
The `?.` is enough here, we set undefined after in either case since it's clearer than doing a conditional